src/cdk/a11y/focus-monitor/focus-monitor.ts

```diff
@@ -388,51 +334,28 @@ export class FocusMonitor implements OnDestroy {
   }
 
   /**
-   * Sets the origin and schedules an async function to clear it at the end of the event queue.
-   * If the detection mode is 'eventual', the origin is never cleared.
+   * Updates the focus origin. If we're using immediate detection mode, we schedule an async
+   * function to clear the origin at the end of a timeout. The duration of the timeout depends on
+   * the origin being set.
    * @param origin The origin to set.
    */
-  private _setOriginForCurrentEventQueue(origin: FocusOrigin): void {
+  private _setOrigin(origin: FocusOrigin): void {
     this._ngZone.runOutsideAngular(() => {
       this._origin = origin;
 
+      // If we're in IMMEDIATE mode, reset the origin at the next tick (or in `TOUCH_BUFFER_MS` ms
+      // for a touch event). We reset the origin at the next tick because Firefox focuses one tick
+      // after the interaction event. We wait `TOUCH_BUFFER_MS` ms before resetting the origin for
+      // a touch event because when a touch event is fired, the associated focus event isn't yet in
+      // the event queue. Before doing so, clear any pending timeouts.
       if (this._detectionMode === FocusMonitorDetectionMode.IMMEDIATE) {
-        // Sometimes the focus origin won't be valid in Firefox because Firefox seems to focus *one*
-        // tick after the interaction event fired. To ensure the focus origin is always correct,
-        // the focus origin will be determined at the beginning of the next tick.
-        this._originTimeoutId = setTimeout(() => this._origin = null, 1);
+        clearTimeout(this._originTimeoutId);
+        const ms = (origin === 'touch') ? TOUCH_BUFFER_MS : 1;
+        this._originTimeoutId = setTimeout(() => this._origin = null, ms);
```

Comment:
Previously, in cases when we _could_ attribute `touch` immediately, we would have done so, right? With this change, it looks like touch is _always_ delayed by 650ms?

Reply:
If I understand you correctly, that's almost correct. Previously, in the case where we could attribute touch immediately (i.e. via `focusVia(el, 'touch'`), we'd immediately update the origin and reset it at the start of the next tick. Now, we still immediately update the origin, but we reset it _650ms later_, instead of at the start of the next tick. Thus, this change does slightly change the behavior of `focusVia` in that regard. It's easy to fix this, if we want, I think it'd be easy to do.

Reply:
I was just trying to make sure I understood (which I did not). Now I follow, I think this makes sense.

Reply:
Okay. I added a `isFromInteractionEvent` parameter to this method that allows us to distinguish from when it is called from `focusVia` versus from the `InputModalityService`. Please LMK if you have a better way of distinguishing between the two.